lib/__tests__.js

```diff
@@ -138,80 +104,995 @@ Test('component API', (t) => {
+  t.notOk(errors, 'no errors');
+  t.end();
+});
+
+Test('error from delegatee propagated back to delegator and abstracted (looks like it came from resolver that called delegate)', async (t) => {
+  const primitive = new GraphQLComponent({
+    types: `
+      type Query {
+        foo: Foo
+      }
+
+      type Foo {
+        b: String
+      }
+    `,
+    resolvers: {
+      Query: {
+        foo() {
+          throw new Error('db retrieval error');
+        }
+      }
+    }
+  });
+
+  const composite = new GraphQLComponent({
+    types: `
+
+      type Query {
+        bar: Bar
+      }
+
+      type Bar {
+        a: String
+        b: String
+      }
+    `,
+    resolvers: {
+      Query: {
+        bar(root, args, context, info) {
+          return GraphQLComponent.delegateToComponent(primitive, {
+            query: 'operation',
+            fieldName: 'foo',
+            context,
+            info
+          });
+        }
+      }
+    }
+  });
+
+  const document = gql`
+    query {
+      bar {
+        a
+        b
+      }
+    }
+  `;
+
+  const { data, errors } = await graphql.execute({
+    schema: composite.schema,
+    document,
+    contextValue: {}
+  });
+
+  t.notOk(data.bar, 'bar is null as expected');
+  t.equals(errors.length, 1, '1 error as expected');
+  t.equals(errors[0].message, 'db retrieval error');
+  t.deepEqual(errors[0].path, ['bar'], 'error path appears as though error came from delegateToComponent calling resolver');
+  t.end();
+});
+
+Test('delegateToComponent maintains backwards compatibility for changed option keys (contextValue and targetRootField)', async (t) => {
+  const primitive = new GraphQLComponent({
+    types: `
+      type Query {
+        foo: Foo
+      }
+
+      type Foo {
+        a: String
+      }
+    `,
+    resolvers: {
+      Query: {
+        foo() {
+          return { a: 'a' };
+        }
+      }
+    }
+  });
+
+  const composite = new GraphQLComponent({
+    types: `
+      type Query {
+        bar: Foo
+      }
+
+      type Foo {
+        b: Int
+      }
+    `,
+    resolvers: {
+      Query: {
+        async bar(root, args, context, info) {
+          const subFoo = await GraphQLComponent.delegateToComponent(primitive, {
+            operation: 'query',
+            targetRootField: 'foo',
+            contextValue: context,
+            info
+          });
+
+          return { ...subFoo, b: 1 };
+        }
+      }
+    },
+    imports: [primitive]
+  });
+
+  const document = gql`
+    query {
+      bar {
+        a
+        b
+      }
+    }
+  `;
+
+  const { data, errors } = await graphql.execute({
+    schema: composite.schema,
+    document,
+    contextValue: {}
+  });
+
+  t.deepEqual(data, { bar: { a: 'a', b: 1}}, 'expected result');
+  t.notOk(errors, 'no errors')
+  t.end();
+});
+
+// directive tests
+
+// TODO: implement directive tests to test directive behavior between imports, etc
```

Comment:
think we might need to write tests here and/or make minor tweaks at schema construction time to make sure directives work the way we think/want when there is a tree of components (ie. imports)

* how do they stack if 2 different directives are added to the same field on the same type defined in 2 components, one who imports another
* conflicts - both the importing and importee components define and use a directive of the same name.